File: 1255-Maximum-Score-Words-Formed-by-Letters/main.py

from collections import Counter
from typing import List

class Solution:
    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:

        letter_count = Counter(letters)
        points = {word: sum([score[ord(char) - ord('a')] for char in word]) for word in words }

        def dfs(i: int) -> int:
            if i == len(words):
                return 0

            word = words[i]
            count = Counter(word)
            if all([letter_count[char] >= count[char] for char in count]):
                for char in count:
                    letter_count[char] -= count[char]
                score = points[word] + dfs(i + 1)
                for char in count:
                    letter_count[char] += count[char]
                
                return max(score, dfs(i + 1))

            return dfs(i+1)

        return dfs(0)

    

File: 1255-maximum-score-words-formed-by-letters/main.go
package main

func maxScoreWords(words []string, letters []byte, score []int) int {

	var scoreMap = make(map[byte]int)
	for i, s := range score {
		scoreMap[byte(i+97)] = s
	}

	var letterMap = make(map[byte]int)
	for _, l := range letters {
		letterMap[l] += 1
	}

	var wordMap = make(map[string]map[byte]int)
	for _, word := range words {
		var m = make(map[byte]int)
		for _, l := range word {
			m[byte(l)] += 1
		}
		wordMap[word] = m
	}

	var maxScore int = 0

	var backtrack func(start int, score int)
	backtrack = func(start int, score int) {

		if score > maxScore {
			maxScore = score
		}

		for i := start; i < len(words); i++ {
			var canAdd bool = true
			var newScore int = score

			for k, v := range wordMap[words[i]] {
				if letterMap[k] < v {
					canAdd = false
					break
				} else {
					newScore += v * scoreMap[k]
				}
			}

			if canAdd {
				for k, v := range wordMap[words[i]] {
					letterMap[k] -= v
				}
				backtrack(i+1, newScore)
				for k, v := range wordMap[words[i]] {
					letterMap[k] += v
				}
			}
		}
	}

	backtrack(0, 0)
	return maxScore
}
